crates/fuel-core/src/executor.rs

```diff
@@ -631,54 +649,85 @@ where
     fn execute_transaction(
         &self,
         idx: u16,
-        tx: &mut Transaction,
+        tx: MaybeCheckedTransaction,
+        tx_id: &TxId,
         header: &PartialBlockHeader,
         execution_data: &mut ExecutionData,
         execution_kind: ExecutionKind,
         tx_db_transaction: &mut DatabaseTransaction,
         options: ExecutionOptions,
-    ) -> ExecutorResult<()> {
-        let tx_id = tx.id(&self.config.transaction_parameters.chain_id);
+    ) -> ExecutorResult<Transaction> {
         // Throw a clear error if the transaction id is a duplicate
         if tx_db_transaction
             .deref_mut()
             .storage::<Transactions>()
-            .contains_key(&tx_id)?
+            .contains_key(tx_id)?
         {
-            return Err(ExecutorError::TransactionIdCollision(tx_id))
+            return Err(ExecutorError::TransactionIdCollision(*tx_id))
         }
 
+        let block_height = *header.height();
         match tx {
```

Comment:
Maybe we should try to reduce some of the duplication in these match arms?

For example, we could first turn a MaybeCheckedTransaction into a CheckedTransaction, and then separately execute the normalized CheckedTransaction.